File: ecom-backend/src/main/java/com/example/demo/Controller/UserController.java
package com.example.demo.Controller;

import com.example.demo.DTO.LoginDTO;
import com.example.demo.DTO.SignUpDTO;
import com.example.demo.Model.User;
import com.example.demo.Response.CustomResponse;
import com.example.demo.Service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/user/")
@CrossOrigin("*")
public class UserController {

    @Autowired
    private UserService userService;


    @PostMapping("/signup")
    @PreAuthorize("hasRole('USER')")
    public CustomResponse createUser(@RequestBody SignUpDTO signUpDTO){
        return userService.registerUser(signUpDTO,false);
    }

    @PostMapping("/login")
    @PreAuthorize("hasRole('USER')")
    public CustomResponse userLogin(@RequestBody LoginDTO loginDTO){
        return userService.userLogin(loginDTO);
    }

    @GetMapping("/{userId}")
    public CustomResponse getUserById(@PathVariable String userId){
        return userService.getUserById(userId);
    }

    @PutMapping("update/{userId}")
    public CustomResponse updateUser(@PathVariable String userId,@RequestBody SignUpDTO signUpDTO){
        return userService.updateUserDetails(userId,signUpDTO);
    }
}

File: bank/src/main/kotlin/com/junocoding/bank/controllers/userController.kt
package com.junocoding.bank.controllers

import com.junocoding.bank.DTOS.userLogin
import com.junocoding.bank.DTOS.userRegister
import com.junocoding.bank.models.User
import com.junocoding.bank.service.userService
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import java.time.LocalDate
import java.util.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Controller
@RestController
@RequestMapping("/user")
@CrossOrigin("*")
class userController(private val userService: userService) {
    @PostMapping("/signup") //api for user signup
    fun register(@RequestBody body: userRegister) : ResponseEntity<User> {
        val user= User()
        user.name=body.name
        user.email=body.email
        user.password=body.password
        user.date= LocalDate.now()
        println("user registration successful")
        return ResponseEntity.ok(userService.save(user))

    }
    @PostMapping("/login")  //api for user login
    fun login(@RequestBody body:userLogin, response: HttpServletResponse):ResponseEntity<Any> {

        val existeduser = userService.findByEmail(body.email)
            ?: return ResponseEntity.badRequest().body("user not found")
        return if (existeduser.password!=body.password) {
            ResponseEntity.badRequest().body("Invalid password")
        } else {
            println("user logged in")
            ResponseEntity.ok(existeduser)
        }

    }
    @PostMapping("/logout")  //api for user logout
    fun logout(request:HttpServletRequest):ResponseEntity<String>{
        request.session.invalidate()
        return ResponseEntity.ok("User logout successfully")
    }
    @GetMapping("/{id}")  //api for getting user
    fun GetUserDetails(@PathVariable id:Long): Optional<User> {
        val existedUser=userService.findUser(id)
        return existedUser
    }
    @PutMapping("/{id}/update")
    fun updateUser(@PathVariable id: Long,@RequestBody body:userRegister): User {
        val updateduser=userService.updateUser(id,body)
        return updateduser
    }

}


